File: kernel/syscall/process.c
// Syscalls for process management

#include <menix/fs/vfs.h>
#include <menix/memory/vm.h>
#include <menix/syscall/syscall.h>
#include <menix/system/abi.h>
#include <menix/system/arch.h>
#include <menix/system/sch/process.h>
#include <menix/system/sch/scheduler.h>
#include <menix/system/sch/thread.h>

#include <uapi/errno.h>

// Forks a thread by cloning its attributes.
SYSCALL_IMPL(fork)
{
	Thread* thread = arch_current_cpu()->thread;
	return SYSCALL_OK(proc_fork(thread->parent, thread));
}

// Terminates the current process.
// `status`: The status code to return to the parent process.
SYSCALL_IMPL(exit, int status)
{
	Process* process = arch_current_cpu()->thread->parent;
	process->return_code = status;
	proc_kill(process, false);
	return SYSCALL_ERR(EFAULT);
}

// Forcefully terminates a process.
// `pid`: The ID of the process to kill.
// `sig`: The signal to send to the process.
SYSCALL_IMPL(kill, usize pid, usize sig)
{
	Process* process = sch_id_to_process(pid);
	if (process == NULL)
		return SYSCALL_ERR(EINVAL);

	// TODO: process->return_code = SIGKILL;
	proc_kill(process, false);
	return SYSCALL_OK(0);
}

// Starts a new process from an ELF executable. Returns 0 upon success, otherwise -1.
// `path`: The path where the executable to load is stored.
// `argv`: A NULL-terminated list of program arguments to be passed to the new process.
// `envp`: A NULL-terminated list of environment variables to be passed to the new process.
SYSCALL_IMPL(execve, const char* path, char** argv, char** envp)
{
	if (path == NULL)
		return SYSCALL_ERR(ENOENT);

	if (proc_create_elf(NULL, path, argv, envp, true) == true)
		return SYSCALL_OK(0);
	else
		return SYSCALL_ERR(arch_current_cpu()->thread->errno);
}

// Returns the ID of the calling process.
SYSCALL_IMPL(getpid)
{
	return SYSCALL_OK(arch_current_cpu()->thread->parent->id);
}

SYSCALL_IMPL(getcwd, VirtAddr buf, usize size)
{
	if (buf == 0 || size == 0)
		return SYSCALL_ERR(EINVAL);
	if (size > PATH_MAX)
		return SYSCALL_ERR(ERANGE);

	Process* proc = arch_current_cpu()->thread->parent;

	// Get the path.
	char* kernel_buf = kmalloc(PATH_MAX);
	usize written = vfs_get_path(proc->working_dir, kernel_buf, size);

	// Copy the result to the user.
	vm_user_write(proc, buf, kernel_buf, written);
	kfree(kernel_buf);

	return SYSCALL_OK(0);
}

SYSCALL_STUB(setuid)
SYSCALL_STUB(getuid)

SYSCALL_STUB(setgid)
SYSCALL_STUB(getgid)

SYSCALL_IMPL(waitpid, pid_t pid, VirtAddr status, int flags)
{
	// TODO
	return SYSCALL_OK(0);
}

SYSCALL_IMPL(sigsuspend)
{
	// TODO: sigmask
	Thread* thread = arch_current_cpu()->thread;
	thread->state = ThreadState_Sleeping;
	while (thread->state == ThreadState_Sleeping)
	{
		sch_arch_invoke();
	}
	return SYSCALL_ERR(EINTR);
}

SYSCALL_STUB(sigprocmask)
SYSCALL_STUB(sigpending)
SYSCALL_STUB(sigaction)
SYSCALL_STUB(sigreturn)
SYSCALL_STUB(sigtimedwait)

SYSCALL_STUB(futex_wait)
SYSCALL_STUB(futex_wake)
SYSCALL_IMPL(gettid)
{
	const usize val = arch_current_cpu()->thread->id;
	return SYSCALL_OK(val);
}

File: kernel/src/syscall/process.rs
use crate::{
    arch::sched::Context,
    {
        memory::{VirtAddr, user::UserPtr},
        percpu::CpuData,
        posix::errno::{EResult, Errno},
        process::ProcessState,
        sched::Scheduler,
        vfs::{File, file::OpenFlags, inode::Mode},
    },
};
use alloc::vec::Vec;
use core::ffi::{CStr, c_char};

pub fn gettid() -> usize {
    Scheduler::get_current().get_id()
}

pub fn getpid() -> usize {
    Scheduler::get_current().get_process().get_pid()
}

pub fn getppid() -> usize {
    Scheduler::get_current()
        .get_process()
        .get_parent()
        .map_or(0, |x| x.get_pid())
}

pub fn getuid() -> usize {
    let proc = Scheduler::get_current().get_process();
    proc.identity.lock().user_id as usize
}

pub fn geteuid() -> usize {
    let proc = Scheduler::get_current().get_process();
    proc.identity.lock().effective_user_id as usize
}

pub fn getgid() -> usize {
    let proc = Scheduler::get_current().get_process();
    proc.identity.lock().group_id as usize
}

pub fn getegid() -> usize {
    let proc = Scheduler::get_current().get_process();
    proc.identity.lock().effective_group_id as usize
}

pub fn getpgid(pid: usize) -> EResult<usize> {
    if pid != 0 {
        return Err(Errno::EINVAL);
    }

    let proc = Scheduler::get_current().get_process();
    Ok(proc.get_pid())
}

pub fn exit(error: usize) -> ! {
    let proc = Scheduler::get_current().get_process();

    if proc.get_pid() <= 1 {
        panic!("Attempted to kill init with error code {error}");
    }

    proc.exit(error as _);
    unreachable!();
}

pub fn fork(ctx: &Context) -> EResult<usize> {
    let old = Scheduler::get_current().get_process();

    // Fork the current process. This puts both processes at this point in code.
    let (new_proc, new_task) = old.fork(ctx)?;
    Scheduler::add_task_to_best_cpu(new_task.clone());

    Ok(new_proc.get_pid())
}

pub fn execve(path: VirtAddr, argv: VirtAddr, envp: VirtAddr) -> EResult<usize> {
    let proc = Scheduler::get_current().get_process();

    let path_str = unsafe { CStr::from_ptr(path.as_ptr()) };

    let args: Vec<_> = (0..)
        .map(|i| unsafe { argv.as_ptr::<usize>().offset(i).read() })
        .take_while(|&p| p != 0)
        .map(|p| {
            unsafe { CStr::from_ptr(p as *const c_char) }
                .to_bytes()
                .to_vec()
        })
        .collect();

    let envs: Vec<_> = (0..)
        .map(|i| unsafe { envp.as_ptr::<usize>().offset(i).read() })
        .take_while(|&p| p != 0)
        .map(|p| {
            unsafe { CStr::from_ptr(p as *const c_char) }
                .to_bytes()
                .to_vec()
        })
        .collect();

    let file = File::open(
        proc.root_dir.lock().clone(),
        proc.working_dir.lock().clone(),
        path_str.to_bytes(),
        OpenFlags::Read | OpenFlags::Executable,
        Mode::empty(),
        &proc.identity.lock(),
    )?;
    proc.fexecve(file, args, envs)?;

    unreachable!("fexecve should never return on success");
}

pub fn waitpid(pid: uapi::pid_t, mut stat_loc: UserPtr<i32>, _options: i32) -> EResult<usize> {
    let proc = Scheduler::get_current().get_process();

    loop {
        let mut inner = proc.children.lock();
        if inner.is_empty() {
            return Err(Errno::ECHILD);
        }
        match pid as isize {
            // Any child process whose process group ID is equal to the absolute value of pid.
            ..=-2 => {
                todo!();
            }
            -1 | 0 => {
                let mut waitee = None;
                for (idx, child) in inner.iter().enumerate() {
                    let child_inner = child.status.lock();
                    if let ProcessState::Exited(code) = *child_inner {
                        stat_loc.write((code as i32) << 8);
                        waitee = Some(idx);
                    }
                }

                if let Some(w) = waitee {
                    inner.remove(w);
                }
            }
            _ => {
                let mut waitee = None;
                for (idx, child) in inner.iter().enumerate() {
                    if child.get_pid() != pid as usize {
                        continue;
                    }

                    let child_inner = child.status.lock();
                    if let ProcessState::Exited(code) = *child_inner {
                        stat_loc.write((code as i32) << 8);
                        waitee = Some(idx);
                    }
                }

                if let Some(w) = waitee {
                    inner.remove(w);
                }
            }
        }
        drop(inner);
        CpuData::get().scheduler.reschedule();
    }
}
